Source/CarthageKit/VersionFile.swift

```diff
@@ -0,0 +1,102 @@
+//
+//  VersionFile.swift
+//  Carthage
+//
+//  Created by Jason Boyle on 8/11/16.
+//  Copyright © 2016 Carthage. All rights reserved.
+//
+
+import Foundation
+
+private let commitishKeyName = "commitish"
+private let cachedFrameworksKeyName = "cachedFrameworks"
+private let frameworkNameKeyName = "frameworkName"
+private let frameworkSha1KeyName = "frameworkSha1"
+
+private typealias VersionFile = [String: [String: AnyObject]]
+
+public func createVersionFileForDependency(dependency: Dependency<PinnedVersion>, forPlatforms platforms: Set<Platform>, buildProductURLs: [NSURL], rootDirectoryURL: NSURL) -> Bool {
+	var cachedPlatformFrameworks: [String: [[String: String]]] = [:]
+	
+	for url in buildProductURLs {
+		guard let platformName = url.URLByDeletingLastPathComponent?.lastPathComponent else { return false }
+		guard let frameworkName = url.URLByDeletingPathExtension?.lastPathComponent else { return false }
+		
+		let frameworkURL = url.appendingPathComponent(frameworkName, isDirectory: false)
+		guard NSFileManager.defaultManager().fileExistsAtPath(frameworkURL.path!) else { return false }
+		guard let sha1 = sha1ForFileAtURL(frameworkURL) else { return false }
+		
+		let frameworkInfo = [frameworkNameKeyName: frameworkName, frameworkSha1KeyName: sha1]
+		
+		var frameworks = cachedPlatformFrameworks[platformName] ?? []
+		frameworks.append(frameworkInfo)
+		cachedPlatformFrameworks[platformName] = frameworks
+	}
+	
+	var versionInfo: VersionFile = [:]
+	
+	let platformSet = platforms.isEmpty ? Set(Platform.supportedPlatforms) : platforms
+	for platform in platformSet {
+		let cachedFrameworks: [[String: String]] = cachedPlatformFrameworks[platform.rawValue] ?? []
+		versionInfo[platform.rawValue] = [commitishKeyName: dependency.version.commitish, cachedFrameworksKeyName: cachedFrameworks]
+	}
+	
+	let rootBinariesURL = rootDirectoryURL.appendingPathComponent(CarthageBinariesFolderPath, isDirectory: true).URLByResolvingSymlinksInPath!
+	let versionFileURL = rootBinariesURL.appendingPathComponent(".\(dependency.project.name).version")
+	
+	if let oldVersionFile: VersionFile = readVersionFileAtURL(versionFileURL) {
+		for platformName in oldVersionFile.keys {
+			if versionInfo[platformName] == nil {
+				versionInfo[platformName] = oldVersionFile[platformName]
+			}
+		}
+	}
+
+	do {
+		let jsonData = try NSJSONSerialization.dataWithJSONObject(versionInfo, options: .PrettyPrinted)
+		try jsonData.writeToURL(versionFileURL, options: .DataWritingAtomic)
+	}
+	catch { return false }
+	
+	return true
+}
+
+public func versionFileMatchesDependency(dependency: Dependency<PinnedVersion>, forPlatforms platforms: Set<Platform>, rootDirectoryURL: NSURL) -> Bool {
```

Comment:
There are no comparisons between version files in the current code; a version file matches if the commitish matches and the recorded SHA1s match the SHA1s of the corresponding frameworks in the Build directory for a given platform.